cpp/src/strings/capitalize.cu

```diff
@@ -0,0 +1,343 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/strings/strings_column_view.hpp>
+#include <cudf/strings/string_view.cuh>
+#include <cudf/strings/case.hpp>
+#include <cudf/utilities/error.hpp>
+#include <strings/char_types/is_flags.h>
+#include <strings/utilities.hpp>
+#include <strings/utilities.cuh>
+
+
+namespace cudf
+{
+namespace strings
+{
+namespace detail
+{
+namespace { // anonym.
+
+  //base class for probing string
+  //manipulation memory load requirements;
+  //and for executing string modification:
+  //
+  struct probe_execute_base
+  {
+    using char_info = thrust::pair<uint32_t,detail::character_flags_table_type>;
+    
+    probe_execute_base(column_device_view const d_column,
```

Comment:
The `d_flags` and `d_case_table` are only needed by the base class now and can retrieve these itself rather than having the subclass pass them through. Something like this:
```
    probe_execute_base(column_device_view const d_column,
                       int32_t const* d_offsets = nullptr,
                       char* d_chars = nullptr):
      d_column_(d_column),
      d_flags_(get_character_flags_table()),      // set flag table
      d_case_table_(get_character_cases_table()), // set case table
      d_offsets_(d_offsets),
      d_chars_(d_chars)
```
Now the subclasses do not have to accept these and pass them through.
```
   execute_capitalize(column_device_view const d_column, int32_t const* d_offsets, char* d_chars)
    : probe_execute_base(d_column, d_offsets, d_chars)
    
```

Reply:
Did you mean `probe_capitalize` in your example above? Because `execute_capitalize` needs all 5 parameters to pass through.

Reply:
No. The `probe_execute_base` does not need to be passed the flag and case tables. It can retrieve them itself as shown in the example code above. Therefore `execute_capitalize` or any other subclass would no longer need to accept or to pass these through.

Reply:
I see now what you're saying. Nice. I fixed that, but I cannot check in: GitHub issues, see my entry in devops channel (internal, only)

Reply:
Okay, looks like I finally was able to check in. Resolving.